Task1.Generate a reproducible sensor matrix
Create a 2D NumPy array named readings with shape (360, 4) to represent six hours of minute-level readings from four sensors. Use a fixed seed and generate values that roughly follow a normal distribution around 50 with some variation. After creating the array, print its shape, dtype, and the first three rows to confirm the structure.

In [23]:
import numpy as np
np.random.seed(42)

In [24]:
rows=360
cols=4
mean = 50       
std_dev = 15      
readings = np.random.normal(loc=mean, scale=std_dev, size=(rows, cols))


In [25]:
print("Shape:", readings.shape)      
print("Dtype:", readings.dtype)      
print("First 3 rows:\n", readings[:3]) 

Shape: (360, 4)
Dtype: float64
First 3 rows:
 [[57.4507123  47.92603548 59.71532807 72.84544785]
 [46.48769938 46.48794565 73.68819223 61.51152094]
 [42.95788421 58.13840065 43.04873461 43.0140537 ]]


Shape: (360, 4)
Dtype: float64
First 3 rows:
 [[51.25730221 48.67895137 56.4042265  51.04900117]
 [44.64330627 53.61595055 63.04000045 59.47080963]
 [42.96264764 37.34578529 43.76725537 50.41325979]]


#### Task 3: Filter outliers with boolean masks
Define outliers as any reading above 80 or below 20. Build a boolean mask that identifies outliers across the full matrix. Use the mask to compute two values: the count of outlier readings and the percentage of total readings that are outliers. Then create a cleaned array cleaned_readings where outliers are replaced with np.nan. Confirm that the number of np.nan values matches your outlier count.

In [26]:
outlier_mask = (readings > 80) | (readings < 20)

In [27]:
outlier_count = np.sum(outlier_mask)
print("Outlier count:", outlier_count)

Outlier count: 65


Outlier count: 56


In [28]:
total_readings=readings.size
outlier_percentage=(outlier_count/total_readings)*100
print("Outlier percentage:", outlier_percentage, "%")

Outlier percentage: 4.513888888888888 %


In [29]:
cleaned_readings = readings.copy()
cleaned_readings[outlier_mask] = np.nan

In [30]:
nan_count = np.isnan(cleaned_readings).sum()
print("NaN count:", nan_count)
print("Counts match:", nan_count == outlier_count)

NaN count: 65
Counts match: True


NaN count: 56
Counts match: True

#### Task 4: Compute summaries by sensor and over time
Compute the mean and standard deviation for each sensor using axis-based aggregation. Store the results in two 1D arrays named sensor_means and sensor_stds. Then compute the average reading at each time step across sensors and store it in a 1D array named time_means. Validate shapes: the sensor summaries should have length 4, and the time summary should have length 360.

In [31]:
sensor_means = np.mean(readings, axis=0)
sensor_stds = np.std(readings, axis=0)

print("Sensor means:", sensor_means)
print("Sensor stds:", sensor_stds)

Sensor means: [50.43849455 50.22541801 50.83432074 50.89241994]
Sensor stds: [13.9980601  15.41966798 14.79864467 14.95712768]


Sensor means: [50.2921572  51.68408191 50.0236014  49.59378015]
Sensor stds: [14.01070979 14.17641004 15.170803   15.14529985]

In [32]:
time_means = np.mean(readings, axis=1)
print("Time means shape:", time_means.shape)

Time means shape: (360,)


In [33]:
print("Sensor means shape:", sensor_means.shape)
print("Sensor stds shape:", sensor_stds.shape)
print("Time means shape:", time_means.shape)

Sensor means shape: (4,)
Sensor stds shape: (4,)
Time means shape: (360,)


#### Task 5: Build a validation report (Part 1)
Create a short report dictionary with keys total_readings, outlier_count, outlier_percent, sensor_means, and sensor_stds. Convert this report into a readable string and print it in the notebook. Include at least one simple validation, such as confirming that total_readings equals readings.size.

In [34]:
report = {
    "total_readings": readings.size,
    "outlier_count": outlier_count,
    "outlier_percent": outlier_percentage,
    "sensor_means": sensor_means,
    "sensor_stds": sensor_stds
}

In [40]:
validation_check = (report["total_readings"] == readings.size)
report["size_validation"] = validation_check
print(validation_check)

True


In [43]:
report_str=f"""
Total readings: {report['total_readings']}
Outlier count: {report['outlier_count']}
Outlier percentage: {report['outlier_percent']:.2f} %

Sensor means: {report['sensor_means']}
Sensor std deviations: {report['sensor_stds']}

Size validation passed: {report['size_validation']}
"""
print(report_str)


Total readings: 1440
Outlier count: 65
Outlier percentage: 4.51 %

Sensor means: [50.43849455 50.22541801 50.83432074 50.89241994]
Sensor std deviations: [13.9980601  15.41966798 14.79864467 14.95712768]

Size validation passed: True



Total readings: 1440
Outlier count: 65
Outlier percentage: 4.51 %

Sensor means: [50.43849455 50.22541801 50.83432074 50.89241994]
Sensor std deviations: [13.9980601  15.41966798 14.79864467 14.95712768]

Size validation passed: True

#### Task 2: Apply vectorized transformations
Create a new array scaled_readings that rescales the original readings to a 0–1 range using vectorized operations. Do not use loops. Then create a second array centered_readings by subtracting the column means from the original readings. Verify both arrays have the same shape as the original.

In [46]:
min_val = np.min(readings)
max_val = np.max(readings)
scaled_readings = (readings - min_val) / (max_val - min_val)
scaled_readings.shape

(360, 4)

In [47]:
column_means = np.mean(readings, axis=0)
centered_readings = readings - column_means
centered_readings.shape

(360, 4)

In [48]:
print("Original shape:", readings.shape)
print("Scaled shape:", scaled_readings.shape)
print("Centered shape:", centered_readings.shape)

Original shape: (360, 4)
Scaled shape: (360, 4)
Centered shape: (360, 4)


Original shape: (360, 4)
Scaled shape: (360, 4)
Centered shape: (360, 4)